lib/posix/pthread.c

```diff
@@ -436,24 +433,26 @@ int pthread_join(pthread_t thread, void **status)
 		return ESRCH;
 	}
 
-	pthread_mutex_lock(&pthread->state_lock);
-
-	if (pthread->state == PTHREAD_JOINABLE) {
-		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
+	if (pthread->state == PTHREAD_DETACHED) {
+		return EINVAL;
 	}
 
+	k_thread_join(&pthread->thread, K_FOREVER);
+
+	pthread_mutex_lock(&pthread->state_lock);
+
 	if (pthread->state == PTHREAD_EXITED) {
+		pthread->state = PTHREAD_TERMINATED;
 		if (status != NULL) {
 			*status = pthread->retval;
 		}
-	} else if (pthread->state == PTHREAD_DETACHED) {
-		ret = EINVAL;
 	} else {
 		ret = ESRCH;
 	}
 
 	pthread_mutex_unlock(&pthread->state_lock);
-	if (pthread->state == PTHREAD_EXITED) {
+
+	if (ret == 0 && pthread->state == PTHREAD_TERMINATED) {
```

Comment:
I think this check is protecting the mutex destroy from happening twice because the thread was already TERMINATED, right? Might add a comment to make it clear, or even use a separate local boolean set in the code above which sets the state.

Reply:
Yes. A detached thread must destroy its own resources in exit, while a joinable thread must destroy resources in join.

Reply:
Added a comment.